File: arbscan/review.py
"""Interactive review of matcher candidates: read both rulebooks, then approve or reject.

Approved pairs are appended to pairs.csv (which the scanner watches); every decision
is remembered so a candidate is only shown once.
"""

import shutil
import sqlite3
import sys
import textwrap
import time
from datetime import datetime, timezone

from .config import Config
from .pairs import append_pair, load_pairs

INVERSE_WARNING = (
    "INVERSE pair: Polymarket YES is treated as Kalshi NO. That only hedges if the "
    "event cannot end any other way: no draw/tie, and both venues handle "
    "postponement or cancellation the same way. Check both rulebooks."
)


def _fmt_ts(ts: float | None) -> str:
    if ts is None:
        return "?"
    return datetime.fromtimestamp(ts, timezone.utc).strftime("%Y-%m-%d %H:%M UTC")


def _fmt_px(v: float | None) -> str:
    return "--" if v is None else f"{v:.3f}".rstrip("0").rstrip(".")


class _Style:
    def __init__(self, enabled: bool):
        self.on = enabled

    def __call__(self, code: str, text: str) -> str:
        return f"\033[{code}m{text}\033[0m" if self.on else text


def _market_block(style: _Style, venue: str, m: sqlite3.Row, width: int) -> list[str]:
    ind = "    "
    if venue == "K":
        head = f"KALSHI   {m['id']}"
        sides = f"YES = {m['yes_label'] or '(see title)'}"
    else:
        head = f"POLY US  {m['id']}"
        sides = f"YES (buy) = {m['yes_label']}    NO (sell) = {m['no_label']}"
    meta = (f"bid/ask {_fmt_px(m['yes_bid'])}/{_fmt_px(m['yes_ask'])}   taker fee coef {m['fee_coef']:.4f}   "
            f"starts {_fmt_ts(m['start_ts'])}   resolves ~{_fmt_ts(m['close_ts'])}")
    lines = [style("1", head), ind + m["title"], ind + style("1", sides), ind + meta, ind + "rules:"]
    for para in (m["rules"] or "(none)").split("\n"):
        if para.strip():
            lines += textwrap.wrap(para.strip(), width=width - 8, initial_indent=ind + "  ",
                                   subsequent_indent=ind + "  ")
    return lines


def record(db: sqlite3.Connection, kalshi: str, pm: str, decision: str, source: str) -> None:
    db.execute("INSERT OR REPLACE INTO decisions (kalshi, pm, decision, ts, source) VALUES (?,?,?,?,?)",
               (kalshi, pm, decision, time.time(), source))


def decide(cfg: Config, db: sqlite3.Connection, kalshi: str, pm: str, decision: str,
           source: str = "human", note: str | None = None, commit: bool = True) -> None:
    """Record a review decision; approvals are appended to pairs.csv."""
    if decision not in ("same", "inverse", "reject"):
        raise ValueError(decision)
    record(db, kalshi, pm, decision, source)
    if commit:
        db.commit()
    if decision != "reject":
        if note is None:
            k = db.execute("SELECT title FROM markets WHERE venue = 'K' AND id = ?", (kalshi,)).fetchone()
            note = (k["title"] if k else "")[:80]
        append_pair(cfg.pairs_path, kalshi, pm, decision, note)


def _pending(db: sqlite3.Connection, min_score: float, pairs_path: str) -> list[sqlite3.Row]:
    existing = {(p.kalshi, p.pm) for p in load_pairs(pairs_path)}
    rows = db.execute(
        "SELECT c.* FROM candidates c LEFT JOIN decisions d ON d.kalshi = c.kalshi AND d.pm = c.pm "
        "WHERE d.kalshi IS NULL AND c.score >= ? ORDER BY c.confident DESC, c.score DESC",
        (min_score,),
    ).fetchall()
    return [r for r in rows if (r["kalshi"], r["pm"]) not in existing]


def list_candidates(db: sqlite3.Connection, min_score: float, limit: int, pairs_path: str) -> None:
    rows = _pending(db, min_score, pairs_path)[:limit]
    print(f"{'score':>5}  {'relation':8} {'kalshi_ticker':44} pm_slug")
    for r in rows:
        flag = "" if r["confident"] else " (?)"
        print(f"{r['score']:5.2f}  {r['relation'] + flag:8} {r['kalshi']:44} {r['pm']}")


def run(cfg: Config, db: sqlite3.Connection, min_score: float) -> None:
    pending = _pending(db, min_score, cfg.pairs_path)
    if not pending:
        print("Nothing to review. Run `arbscan match` (or lower --min-score).")
        return
    style = _Style(sys.stdout.isatty())
    width = min(shutil.get_terminal_size((110, 40)).columns, 140)
    approved = rejected = 0
    print(f"{len(pending)} candidates to review, best first. Read both rulebooks before approving.\n")
    for i, c in enumerate(pending, start=1):
        k = db.execute("SELECT * FROM markets WHERE venue = 'K' AND id = ?", (c["kalshi"],)).fetchone()
        p = db.execute("SELECT * FROM markets WHERE venue = 'P' AND id = ?", (c["pm"],)).fetchone()
        if k is None or p is None:
            continue  # catalog refreshed since matching; market gone
        relation = c["relation"]
        meaning = "Polymarket YES = Kalshi YES" if relation == "same" else "Polymarket YES = Kalshi NO"
        confidence = "" if c["confident"] else "  (relation guessed: no outcome labels to compare)"
        print(style("2", "─" * width))
        print(f"[{i}/{len(pending)}]  score {c['score']:.2f}   proposed: "
              f"{style('1;36', relation.upper())} ({meaning}){confidence}")
        print()
        print("\n".join(_market_block(style, "K", k, width)))
        print()
        print("\n".join(_market_block(style, "P", p, width)))
        if relation == "inverse":
            print()
            print(style("1;33", textwrap.fill(INVERSE_WARNING, width=width)))
        print()
        while True:
            try:
                ans = input("[s]ame  [i]nverse  [r]eject  [Enter] skip  [q]uit > ").strip().lower()
            except EOFError:
                ans = "q"
            if ans in ("s", "i", "r", "", "q"):
                break
        if ans == "q":
            break
        if ans == "":
            continue
        decision = {"s": "same", "i": "inverse", "r": "reject"}[ans]
        decide(cfg, db, c["kalshi"], c["pm"], decision)
        if decision == "reject":
            rejected += 1
        else:
            approved += 1
            print(style("32", f"added to {cfg.pairs_path}"))
    print(f"\n{approved} approved, {rejected} rejected. The scanner picks up pairs.csv changes automatically.")

File: arbscan/web/static/js/pages/review.js
import { html, useCallback, useEffect, useRef, useState } from "../vendor/preact-htm.js";
import { api, dateTime, duration, highlight, int, pmTitle, price, splitTitle, toast, usePref, useStore } from "../lib.js";
import { Banner, Card, Empty, Icon, Kbd, RelationChip } from "../ui.js";

const INVERSE_WARNING = html`<b>Inverse pair.</b> Polymarket YES is treated as Kalshi NO. That only hedges if the event
  can't end any other way: no draw or tie, and both venues handle postponement or cancellation the same way.
  Check both rulebooks below.`;

function Score({ value }) {
  return html`<span class="score" title="Match score"><span class="track"><i style=${`width:${Math.min(100, value * 100)}%`}></i></span>${value.toFixed(2)}</span>`;
}

// Which outcome on one venue is the same bet as which on the other.
export function Mapping({ relation, k, p }) {
  const inv = relation === "inverse";
  const kYes = k.yes_label || splitTitle(k.title).tail || "YES";
  return html`<div>
    <div class="mapping" role="img" aria-label=${inv ? "Kalshi YES matches Polymarket NO; Kalshi NO matches Polymarket YES" : "Kalshi YES matches Polymarket YES; Kalshi NO matches Polymarket NO"}>
      <div class="outcome" style="grid-column:1;grid-row:1"><small>Kalshi YES</small>${kYes}</div>
      <div class="outcome" style="grid-column:1;grid-row:2"><small>Kalshi NO</small>Not: ${kYes}</div>
      <svg viewBox="0 0 72 100" preserveAspectRatio="none" style="grid-column:2;grid-row:1 / 3" aria-hidden="true">
        <path class="linkline" vector-effect="non-scaling-stroke" d=${inv ? "M0,23 C36,23 36,77 72,77" : "M0,23 H72"} />
        <path class="linkline alt" vector-effect="non-scaling-stroke" d=${inv ? "M0,77 C36,77 36,23 72,23" : "M0,77 H72"} />
      </svg>
      <div class=${`outcome ${inv ? "" : "hl"}`} style="grid-column:3;grid-row:1"><small>Polymarket US YES · buy</small>${p.yes_label}</div>
      <div class=${`outcome ${inv ? "hl" : ""}`} style="grid-column:3;grid-row:2"><small>Polymarket US NO · sell</small>${p.no_label}</div>
    </div>
    <div class="muted" style="font-size:12.5px;margin-top:8px">
      Lines join the same outcome on both venues. An arb buys opposite outcomes, one on each venue, so exactly one pays $1:
      ${inv ? " Kalshi YES + Poly YES, or Kalshi NO + Poly NO." : " Kalshi YES + Poly NO, or Kalshi NO + Poly YES."}
    </div>
  </div>`;
}

function Rules({ text }) {
  return html`<div class="rules">${highlight(text).map((s) => (s.kind ? html`<mark class=${s.kind}>${s.t}</mark>` : s.t))}</div>`;
}

function Venue({ tag, m, extra }) {
  return html`<div class="venue">
    <div class="venue-tag">${tag}</div>
    <h3>${m.title}</h3>
    <div class="id">${m.id}</div>
    <dl class="facts">
      ${extra}
      <dt>Starts</dt><dd>${m.start_ts ? dateTime(m.start_ts) : "—"}</dd>
      <dt>Resolves</dt><dd>${m.close_ts ? `~${dateTime(m.close_ts)}` : "—"}</dd>
      <dt>YES bid / ask</dt><dd>${price(m.yes_bid)} / ${price(m.yes_ask)} <span class="muted">at catalog time</span></dd>
      <dt>Taker fee</dt><dd>${m.fee_coef?.toFixed(4)} × p(1−p) per contract</dd>
    </dl>
    <${Rules} text=${m.rules || "(no rules text)"} />
  </div>`;
}

export function RulesCompare({ k, p }) {
  const dt = k.start_ts && p.start_ts ? Math.abs(k.start_ts - p.start_ts) : null;
  return html`<div class="grid" style="gap:12px">
    ${dt !== null && dt > 2 * 3600 && html`<${Banner} tone="warning">Start times differ by <b>${duration(dt)}</b>. Make sure these are the same event.<//>`}
    <div class="compare">
      <${Venue} tag="Kalshi" m=${k} extra=${html`<dt>YES</dt><dd>${k.yes_label || "—"}</dd>`} />
      <${Venue} tag="Polymarket US" m=${p} extra=${html`<dt>YES / NO</dt><dd>${p.yes_label} / ${p.no_label}</dd>`} />
    </div>
    <div class="muted" style="font-size:12px">Highlighted: numbers and dates, and wording that changes how a market settles (draws, postponement, exclusions, deadlines).</div>
  </div>`;
}

export function Review() {
  const [q, setQ] = useState("");
  const [minScore, setMinScore] = usePref("review.min", 0.45);
  const [confident, setConfident] = usePref("review.confident", false);
  const [relation, setRelation] = useState("");
  const [items, setItems] = useState([]);
  const [total, setTotal] = useState(0);
  const [sel, setSel] = useState(0);
  const [loading, setLoading] = useState(true);
  const [busy, setBusy] = useState(false);
  const matchUpdated = useStore((s) => s.state?.pipeline?.match?.updated);
  const listRef = useRef();

  const query = useCallback((offset) => {
    const u = new URLSearchParams({ min_score: minScore, offset, limit: 40 });
    if (confident) u.set("confident", "1");
    if (relation) u.set("relation", relation);
    if (q.trim()) u.set("q", q.trim());
    return api(`/api/candidates?${u}`);
  }, [q, minScore, confident, relation]);

  useEffect(() => {
    setLoading(true);
    const t = setTimeout(() => query(0).then((d) => { setItems(d.items); setTotal(d.total); setSel(0); }).finally(() => setLoading(false)), 200);
    return () => clearTimeout(t);
  }, [query, matchUpdated]);

  const more = useCallback(async () => {
    const d = await query(items.length);
    setItems((xs) => [...xs, ...d.items.filter((it) => !xs.some((x) => x.kalshi === it.kalshi && x.pm === it.pm))]);
    setTotal(d.total);
  }, [query, items.length]);

  const cur = items[sel];
  const decide = useCallback(async (decision) => {
    if (!cur || busy) return;
    setBusy(true);
    try {
      await api("/api/candidates/decide", { method: "POST", body: { kalshi: cur.kalshi, pm: cur.pm, decision } });
      toast(decision === "reject" ? "Rejected" : `Approved as ${decision}. Scanning starts next sweep.`);
      setItems((xs) => xs.filter((x) => x !== cur));
      setTotal((t) => t - 1);
      setSel((i) => Math.min(i, items.length - 2));
      if (items.length < 12 && total > items.length) more();
    } finally {
      setBusy(false);
    }
  }, [cur, busy, items.length, total, more]);

  useEffect(() => {
    const fn = (e) => {
      if (e.target.closest("input, select, textarea") || e.metaKey || e.ctrlKey || e.altKey) return;
      const key = e.key.toLowerCase();
      if (key === "s") decide("same");
      else if (key === "i") decide("inverse");
      else if (key === "r") decide("reject");
      else if (key === "j" || key === "arrowdown" || key === "arrowright") setSel((i) => Math.min(items.length - 1, i + 1));
      else if (key === "k" || key === "arrowup" || key === "arrowleft") setSel((i) => Math.max(0, i - 1));
      else return;
      e.preventDefault();
    };
    addEventListener("keydown", fn);
    return () => removeEventListener("keydown", fn);
  }, [decide, items.length]);

  useEffect(() => {
    listRef.current?.querySelector(".q-item.sel")?.scrollIntoView({ block: "nearest" });
  }, [sel]);

  return html`<div class="review">
    <section class="card queue" aria-label="Review queue">
      <div class="queue-filters">
        <label class="search"><${Icon} name="search" size=${15} /><input class="input" placeholder="Search titles or tickers" value=${q} onInput=${(e) => setQ(e.target.value)} aria-label="Search candidates" /></label>
        <div style="display:flex;gap:10px;align-items:center">
          <label class="muted nowrap" style="font-size:12.5px" for="minscore">Min score</label>
          <input id="minscore" type="range" min="0.3" max="1" step="0.05" value=${minScore} onInput=${(e) => setMinScore(+e.target.value)} style="flex:1" />
          <span class="num" style="font-size:12.5px;width:30px">${minScore.toFixed(2)}</span>
        </div>
        <div style="display:flex;gap:12px;align-items:center;flex-wrap:wrap">
          <label class="check"><input type="checkbox" checked=${confident} onChange=${(e) => setConfident(e.target.checked)} />Labels matched</label>
          <select class="select" value=${relation} onChange=${(e) => setRelation(e.target.value)} aria-label="Relation">
            <option value="">Same or inverse</option><option value="same">Same only</option><option value="inverse">Inverse only</option>
          </select>
          <span class="muted num" style="margin-left:auto;font-size:12.5px">${int(total)} pending</span>
        </div>
      </div>
      <div class="queue-list" ref=${listRef} style=${loading ? "opacity:.55" : ""}>
        ${items.map((c, i) => html`<div class=${`q-item ${i === sel ? "sel" : ""}`} key=${c.kalshi + c.pm} onClick=${() => setSel(i)}>
          <div class="row1"><${Score} value=${c.score} /><${RelationChip} relation=${c.relation} /><span class="spacer"></span>
            ${!c.confident && html`<span class="muted" style="font-size:11px" title="No outcome labels to compare">guessed</span>`}</div>
          <div class="k">${splitTitle(c.k?.title).head}${c.k?.yes_label ? html`<span class="muted"> · ${c.k.yes_label}</span>` : ""}</div>
          <div class="p">Poly US · ${pmTitle(c.p?.title)}</div>
        </div>`)}
        ${items.length < total && html`<div style="padding:12px;text-align:center"><button class="btn sm" onClick=${more}>Load more</button></div>`}
        ${!loading && !items.length && html`<${Empty} icon="review" title="Queue is empty">Nothing matches these filters. New suggestions arrive after each refresh.<//>`}
      </div>
    </section>

    ${cur ? html`<section class="card" aria-label="Candidate">
      <div class="card-head">
        <div style="min-width:0">
          <h2 style="font-size:16px">${splitTitle(cur.k.title).head}</h2>
          <div class="sub" style="display:flex;gap:10px;align-items:center;margin-top:6px;flex-wrap:wrap">
            <${Score} value=${cur.score} /><span>Proposed</span><${RelationChip} relation=${cur.relation} />
            <span>${cur.relation === "inverse" ? "Polymarket YES = Kalshi NO" : "Polymarket YES = Kalshi YES"}</span>
          </div>
        </div>
        <div class="actions"><span class="muted num" style="font-size:12px">${sel + 1} of ${int(total)}</span></div>
      </div>
      <div class="card-body grid" style="gap:16px">
        ${!cur.confident && html`<${Banner} tone="info">The relation is a guess: Polymarket gives no outcome labels to compare. Check which side each YES refers to.<//>`}
        ${cur.relation === "inverse" && html`<${Banner} tone="warning">${INVERSE_WARNING}<//>`}
        <${Mapping} relation=${cur.relation} k=${cur.k} p=${cur.p} />
        <${RulesCompare} k=${cur.k} p=${cur.p} />
      </div>
      <div class="actionbar">
        <button class=${`btn ${cur.relation === "same" ? "primary" : ""}`} disabled=${busy} onClick=${() => decide("same")}>
          <${Icon} name="equal" size=${14} />Approve as same <${Kbd}>S<//></button>
        <button class=${`btn ${cur.relation === "inverse" ? "primary" : ""}`} disabled=${busy} onClick=${() => decide("inverse")}>
          <${Icon} name="swap" size=${14} />Approve as inverse <${Kbd}>I<//></button>
        <button class="btn danger" disabled=${busy} onClick=${() => decide("reject")}><${Icon} name="xcircle" size=${14} />Reject <${Kbd}>R<//></button>
        <span class="spacer"></span>
        <button class="btn ghost" onClick=${() => setSel((i) => Math.min(items.length - 1, i + 1))}>Skip <${Kbd}>J<//></button>
      </div>
    </section>` : html`<${Card}>${loading ? html`<div class="muted">Loading…</div>` : html`<${Empty} icon="check" title="All caught up">
      No pending suggestions at this score. Lower the minimum score, or wait for the next refresh.<//>`}<//>`}
  </div>`;
}
